### PR TITLE
refactor: Allow TrackState measurement setting without extra initialization

### PR DESCRIPTION
**Pull request by paulgessinger**
Previously, `VectorMultiTrajectory` would initialize measurement values and covariances to zero. This has some amount of overhead.

With this

1. Calling `allocateCalibrated` on a track state with an existing allocation of different size is now an error
2. An extra overload to `allocateCalibrated` now accepts arguments for the value and covariance that it will initialize the allocation to.
3. The previous signature remains valid, and will initialize the allocation to zero.

## Summary by CodeRabbit

- **New Features**
	- Introduced new methods for allocating and managing calibrated measurements across various classes.
	- Added a `size()` method to `VectorTrackContainer` for easy retrieval of track count.
	- Enhanced `CombinatorialKalmanFilter` with a new `findTracks` method for improved track finding capabilities.
	- Added a new `MultiTrajectoryAllocateCalibratedInit` test case to validate the allocation of calibrated initial states.

- **Bug Fixes**
	- Improved error handling in the `findTracks` method to log propagation errors.
	- Updated memory management in `VectorMultiTrajectory` to ensure proper state invalidation.

- **Tests**
	- Added new test cases for validating the allocation and initialization of calibrated states.
	- Modified existing tests to reflect changes in the handling of calibrated measurements.

### CODE REVIEW

**Comment by paulgessinger** [comment]
I benchmarked this only with the synthetic Track EDM benchmark:

main:
```
Time (mean ± σ):     17.270 s ±  0.115 s    [User: 17.053 s, System: 0.130 s]
Range (min … max):   17.132 s … 17.487 s    10 runs
```

This PR:
```
Time (mean ± σ):     16.553 s ±  0.052 s    [User: 16.352 s, System: 0.125 s]
Range (min … max):   16.488 s … 16.665 s    10 runs
```

File: Core/include/Acts/EventData/VectorMultiTrajectory.hpp [hunk 1]
```diff
@@ -47,6 +47,32 @@ using MultiTrajectoryTraits::IndexType;
 constexpr auto kInvalid = MultiTrajectoryTraits::kInvalid;
 constexpr auto MeasurementSizeMax = MultiTrajectoryTraits::MeasurementSizeMax;
 
+template <typename T>
+struct NonInitializingAllocator {
+  using value_type = T;
```
Comment (andiwand) [inline Core/include/Acts/EventData/VectorMultiTrajectory.hpp:52]: nitpick: should this go in a utility or detail header?

**Review by andiwand — APPROVED** [review]
Looks good to me!

Does this need to be synchronized with an Athena patch?

File: Core/include/Acts/EventData/VectorMultiTrajectory.hpp (hunk 1 above)
Comment (paulgessinger) [reply Core/include/Acts/EventData/VectorMultiTrajectory.hpp:52]: It's currently in the detail namespace for this file, that seemed sufficient from my point of view.

**Comment by paulgessinger** [comment]
@andiwand I don't think so, as the previous API should stay valid, as well as the assumption of memory be zero-initialized. 

We might want to make use of the direct initialization however. 

Also, I'll probably want to time the CKF timing with this, although I don't expect significant improvements globally.

**Comment by paulgessinger** [comment]
I'll also fix the sonar issues tomorrow.

File: Core/include/Acts/EventData/TrackStateProxy.hpp [hunk 2]
```diff
@@ -865,10 +865,33 @@ class TrackStateProxy {
...
+                          const Eigen::DenseBase<cov_t>& cov)
+    requires(Eigen::PlainObjectBase<val_t>::RowsAtCompileTime ==
+                 Eigen::PlainObjectBase<cov_t>::RowsAtCompileTime &&
+             Eigen::PlainObjectBase<cov_t>::RowsAtCompileTime ==
+                 Eigen::PlainObjectBase<cov_t>::ColsAtCompileTime)
+  {
+    m_traj->template allocateCalibrated<
+        Eigen::PlainObjectBase<val_t>::RowsAtCompileTime>(m_istate, val, cov);
+  }
+
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/EventData/TrackStateProxy.hpp:894]: Good point actually.

File: Core/include/Acts/EventData/VectorMultiTrajectory.hpp [hunk 3]
```diff
@@ -464,23 +490,52 @@
...
+    double* measPtr = &m_meas[m_measOffset[istate]];
+    Eigen::Map<ActsVector<measdim>> valMap(measPtr);
+    valMap = val;
+
     m_measCovOffset[istate] = static_cast<IndexType>(m_measCov.size());
     m_measCov.resize(m_measCov.size() + measdim * measdim);
+
+    double* covPtr = &m_measCov[m_measCovOffset[istate]];
+    Eigen::Map<ActsSquareMatrix<measdim>> covMap(covPtr);
+    covMap = cov;
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/EventData/VectorMultiTrajectory.hpp:538]: Makes sense.

File: Core/include/Acts/EventData/VectorMultiTrajectory.hpp [hunk 4]
```diff
@@ -320,9 +346,9 @@
       m_params;
   std::vector<typename detail_lt::FixedSizeTypes<eBoundSize>::Covariance> m_cov;
 
-  std::vector<double> m_meas;
+  std::vector<double, NonInitializingAllocator<double>> m_meas;
   std::vector<MultiTrajectoryTraits::IndexType> m_measOffset;
-  std::vector<double> m_measCov;
+  std::vector<double, NonInitializingAllocator<double>> m_measCov;
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/EventData/VectorMultiTrajectory.hpp:351]: This is ensured by the rest of the interface.

**Comment by paulgessinger** [comment]
There might be a tiny improvement in CKF time, but it's likely at the noise level.

ckf

**Review by andiwand — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [7caafd79a4a22d25367b8856edddc698c696b72e]
### Build job with this PR failed!
Please investigate the [build job](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46542722) for the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/8596538)!

**Pull request merged**